presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java

```diff
@@ -44,12 +47,12 @@ public Pattern getPattern()
 
         WindowNode parent = (WindowNode) node;
 
-        PlanNode source = context.getLookup().resolve(parent.getSource());
-        if (!(source instanceof WindowNode)) {
+        Optional<WindowNode> childOption = pullUnaryNodeAboveProjects(context.getLookup(), WindowNode.class, parent.getSource());
```

Comment:
I think this could be extracted as separate rule, `ProjectionPushDownOverWindoNode`. We should try to avoid doing a recursion in rules.